File: kmeans.c
# include <stdio.h>
# include <stdlib.h>
# include <math.h>

struct cord
{
    double value;
    struct cord *next;
};
struct vector
{
    struct vector *next;
    struct cord *cords;
};

double distance(double *p, double *q, int d)
{
    int i;
    double distance = 0;
    double sum_of_squares = 0;
    for(i=0; i<d; i++)
    {
        sum_of_squares += pow(p[i]-q[i],2);
    }
    distance = sqrt(sum_of_squares);
    return distance;
}

int clusterAssign(double *x, double **centroids, int k, int d)
{
    double min_distance = distance(x, centroids[0],d);
    int cluster = 0;
    int i;
    double dist;
    for(i=1; i<k; i++)
    {
        dist = distance(x, centroids[i],d);
        if(dist <= min_distance)
        {
            min_distance = dist;
            cluster = i;
        }
    }
    return cluster;
}

void printClusters(double **clusters, int d,  int k){
    int i=0;
    int j=0;
    for(i=0; i<k; i++){
        for(j=0; j<d; j++){
            printf("%.4f", clusters[i][j]);
            if(j<d-1){
                printf("%c", ',');
            }
        }
        printf("\n");
    }

}

void centroidUpdate(double **clusters, int *size_of_clusters, int k, int d)
{
    int i = 0;
    int j = 0;
    for(i=0; i < k; i++)
    {   
        for(j=0; j<d ; j++)
        {
            if(size_of_clusters[i]==0){
                printf("An Error Has Occurred\n");
                exit(1);
            }
            clusters[i][j] /= size_of_clusters[i];
        }
    }
    return;
}

void delete_cord( struct cord *head ){
    if ( head != NULL )
        {
            delete_cord( head->next );
            free( head );
        }
    }

void delete_vector(struct vector *head){
    if(head!=NULL){
        delete_vector(head->next);
        delete_cord(head->cords);
        free(head);
    }
}

int main(int argc, char **argv)
{ 
    int k;
    int iter; 
    int num_of_elements=0;
    int d=0;
    int is_end_of_line = 0;
    double *elements_1d;
    double **elements;
    double *centroids_1d;
    double **centroids;
    double *clusters_1d;
    double **clusters;
    int *size_of_clusters;
    int assigned_centroid;
    
    struct vector *head_vec, *curr_vec;
    struct cord *head_cord, *curr_cord;
    int i, j, l;
    double n;
    char c;
    struct vector *loop_vector;
    struct cord *loop_cord;

    int end = 0;

    if (argc == 2) 
    {
        k = atoi(argv[1]);
        iter = 200;
    }
    else if(argc ==3)
    {
        k = atoi(argv[1]);
        iter = atoi(argv[2]);
        if(iter<1 || iter>1000 || iter!= atof(argv[2])){
            printf("Invalid maximum iteration!\n");
            exit(1);
        }
    }
    else
    {
        printf("An Error Has Occurred\n");
        exit(1);
    }

    head_cord = malloc(sizeof(struct cord)); 
    if(head_cord==NULL){
        printf("An Error Has Occurred\n");
        exit(1);
    }
    curr_cord = head_cord;
    curr_cord->next = NULL;

    head_vec = malloc(sizeof(struct vector));
    if(head_vec==NULL){
        printf("An Error Has Occurred\n");
        free(head_cord);
        exit(1);
    }
    curr_vec = head_vec;
    curr_vec->next = NULL;


    while (scanf("%lf%c", &n, &c) == 2)
    {
        if(!is_end_of_line){
            d+=1;
        }
        if (c == '\n')
        {   is_end_of_line =1;
            curr_cord->value = n;
            curr_vec->cords = head_cord;
            curr_vec->next = calloc(1,sizeof(struct vector));
            if(curr_vec->next==NULL){
                printf("An Error Has Occurred\n");
                delete_vector(head_vec);
                exit(1);
            }
            curr_vec = curr_vec->next;
            curr_vec->next = NULL;
            head_cord = calloc(1,sizeof(struct cord));
            if(head_cord==NULL){
                printf("An Error Has Occurred\n");
                delete_vector(head_vec);
                exit(1);
            }
            curr_cord = head_cord;
            curr_cord->next = NULL;
            num_of_elements++;
            continue;
        }

        curr_cord->value = n;
        curr_cord->next = malloc(sizeof(struct cord));
        curr_cord = curr_cord->next;
        curr_cord->next = NULL;
    }
    
    free(curr_cord);

    if(k<1 || k>num_of_elements || k!= atof(argv[1])){
        printf("Invalid number of clusters!\n");
        delete_vector(head_vec);
        exit(1);
    }


    /*memory allocation for all the points in the file*/
    elements_1d = calloc(num_of_elements*d, sizeof(double));
    if(elements_1d == NULL){
        printf("An Error Has Occurred\n");
        delete_vector(head_vec);
        exit(1);
    }
    elements = calloc(num_of_elements,sizeof(double *));
    if(elements == NULL){
        printf("An Error Has Occurred\n");
        delete_vector(head_vec);
        free(elements_1d);
        exit(1);
    }
    for(i=0; i<num_of_elements; i++)
    {
        elements[i] = elements_1d+i*d;
    } 

    /*memory allocation for k centroids*/
    centroids_1d = calloc(k*d, sizeof(double));
    if(centroids_1d == NULL){
        printf("An Error Has Occurred\n");
        delete_vector(head_vec);
        free(elements_1d);
        free(elements);
        exit(1);
    }
    centroids = calloc(k,sizeof(double *));
    if(centroids == NULL){
        printf("An Error Has Occurred\n");
        delete_vector(head_vec);
        free(elements_1d);
        free(elements);
        free(centroids_1d);
        exit(1);
    }
    for(i=0; i<k; i++)
    {
        centroids[i] = centroids_1d+i*d;
    }

    /*put the first k points in the centroids array*/
    i=0;
    j=0;
    
    loop_vector = head_vec;
    for(i=0; i<k; i++){
        loop_cord = loop_vector->cords;
        for(j=0; j<d; j++){
            centroids[i][j]= loop_cord->value;
            loop_cord=loop_cord->next;
        }
        loop_vector=loop_vector->next;
    }
    
    /*memory allocation for sum matrix "representing" clusters*/
    clusters_1d = calloc(k*d, sizeof(double));
    if(clusters_1d == NULL){
        printf("An Error Has Occurred\n");
        delete_vector(head_vec);
        free(elements_1d);
        free(elements);
        free(centroids_1d);
        free(centroids);
        exit(1);
    }
    clusters = calloc(k,sizeof(double *));
    if(clusters == NULL){
        printf("An Error Has Occurred\n");
        delete_vector(head_vec);
        free(elements_1d);
        free(elements);
        free(centroids_1d);
        free(centroids);
        free(clusters_1d);
        exit(1);
    }
    for(j=0; j<k; j++)
    {
        clusters[j] = clusters_1d+j*d;
    } 

    /*memory allocation for array of sizes of clusters*/
    size_of_clusters = calloc(k, sizeof(int));
    if(size_of_clusters == NULL){
        printf("An Error Has Occurred\n");
        delete_vector(head_vec);
        free(elements_1d);
        free(elements);
        free(centroids_1d);
        free(centroids);
        free(clusters_1d);
        free(clusters);
        exit(1);
    }
    
    /*assignment of elements into 2d matrix*/
    i=0;
    j=0;
    loop_vector = head_vec;
    for(i=0; i<num_of_elements; i++){
        loop_cord = loop_vector->cords;
        for(j=0; j<d; j++){
            elements[i][j]= loop_cord->value;
            loop_cord=loop_cord->next;
        }
        loop_vector=loop_vector->next;
    }
    delete_vector(head_vec);
    
    i=0;
    j=0;
    l=0;
    for(i=0; i<iter; i++)
    { 
       for(j=0; j<num_of_elements; j++){ 
        assigned_centroid= clusterAssign(elements[j], centroids,k,d);
        for(l=0; l<d; l++){
            clusters[assigned_centroid][l] += elements[j][l];
        }
        size_of_clusters[assigned_centroid]+=1;
       }
       centroidUpdate(clusters, size_of_clusters, k, d);
       j=0;
       end=1;
       for(j=0; j<k; j++){
        if(distance(centroids[j],clusters[j],d)>= 0.001){
            end = 0;
            break;
        }
       
       }
       if(end){
        break;
       }
       j=0;
       l=0;
       for(j=0;j<k;j++){
        for(l=0;l<d;l++){
            centroids[j][l] = clusters[j][l];
            clusters[j][l]=0;
        }
        size_of_clusters[j]=0;
       }
    }

    
    printClusters(centroids, d, k);

    free(elements);
    free(centroids);
    free(centroids_1d);
    free(elements_1d);
    free(clusters);
    free(size_of_clusters);
    free(clusters_1d);

    
    return 0;
}

File: kmeans.py
import math
import sys
import copy

#p1, p2 are arrays with float values, representing d dimentional points
#the function calculates the euclidean distance between p1 and p2
def distance(p1, p2, d):
    sum_of_squares = 0.0
    for i in range(d):
        sum_of_squares += math.pow(float(p1[i])-float(p2[i]), 2)
    distance= math.sqrt(sum_of_squares)
    return distance

#x is a point in R^d, centroids is an array with k points in R^d
def clusterAssign(x, centriods, k, d):
   min_distance = distance(x, centriods[0], d)
   cluster = 0
   for i in range(1, k):
       dist = distance(x, centriods[i], d)
       if dist <= min_distance:
           min_distance = dist
           cluster = i
   return cluster 

#clusters is a 2D array of sum of the points in each cluster
#size of clusters is an array of the size of each cluster
#k is the number of clusters
#d is the number of coordinates of each point
def centriodUpdate(clusters, size_of_clusters, k: int, d: int):
    for i in range(k):
        for j in range(d):
            if size_of_clusters[i]==0:
                print("An Error Has Occurred")
                exit(0)
            clusters[i][j] =  clusters[i][j]/size_of_clusters[i]
    return

def printClusters(clusters, d, k):
    for i in range(k):
        for j in range(d):
            print('%.4f' % clusters[i][j], end='')
            if j<d-1:
                print(",", end='')
        print('')
    return

def is_float(n):
    try:
        float(n)
        return True
    except:
        return False

#"main"
def main():
    k=0
    iter=0
    file=None
    file_name=""
    elements = []
    centroids = []
    num_of_elements=0
    d=0
    line_count =0
    end=0

    if len(sys.argv) == 3:
        if sys.argv[1].isdigit():
            k = int(sys.argv[1])
        else:
            print("Invalid number of clusters!")
            exit(0)
        iter = 200
        file_name = sys.argv[2] 
    elif len(sys.argv) == 4:
         if sys.argv[1].isdigit():
            k = int(sys.argv[1])
         else:
            print("Invalid number of clusters!")
            exit(0)
         if sys.argv[2].isdigit():
             iter = int(sys.argv[2])
         else:
            print("Invalid maximum iteration!")
            exit(0)
         file_name = str(sys.argv[3])
         if iter<1 or iter>1000:
            print("Invalid maximum iteration!")
            exit(0)
    else:
        print("An Error Has Occurred")
        exit(0)

    file = open(file_name, "r")
    for line in file:
        point = [float(n) for n in line.split(',') if is_float(n)]
        elements.append(point)
        line_count+=1
        if line_count<=k:
            centroids.append(copy.deepcopy(point))

    num_of_elements=len(elements)
    d=len(elements[0])

    if k<1 or k>num_of_elements:
        print("Invalid number of clusters!")
        exit(0)
    

    clusters = [[0.0 for i in range(d)] for j in range(k)] 
    size_of_clusters = [0 for i in range(k)]

    assigned_centriod=0
    for i in range(iter):
        for j in range(num_of_elements):
            assigned_centriod = clusterAssign(elements[j],centroids, k, d)
            for l in range(d):
                clusters[assigned_centriod][l] += elements[j][l]
            size_of_clusters[assigned_centriod]+=1
        centriodUpdate(clusters, size_of_clusters, k, d)
        j=0
        end=1
        for j in range(k):
            if distance(centroids[j], clusters[j], d)>=0.001:
                end=0
                break
        if end:
            break  
        j=0
        l=0
        for j in range(k):
            for l in range(d):
                centroids[j][l] = float(clusters[j][l])
                clusters[j][l] = 0.0
            size_of_clusters[j] = 0

    printClusters(centroids, d, k)

if __name__ == "__main__":
    main()